File: cpp/src/0021.cpp
/*
[21. Merge Two Sorted Lists (Easy)](https://leetcode.com/problems/merge-two-sorted-lists/)

@topic: Linked List
@topic: Recursion

Constraints:
  - The number of nodes in both lists is in the range `[0, 50]`.
  - `-100 <= Node.val <= 100`
  - Both `list1` and `list2` are sorted in **non-decreasing** order.
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        auto fake = ListNode();
        auto curr = &fake;
        while (list1 && list2) {
            if (list1->val <= list2->val) {
                curr->next = list1;
                list1 = list1->next;
            } else {
                curr->next = list2;
                list2 = list2->next;
            }
            curr = curr->next;
        }
        curr->next = list1 ? list1 : list2;
        return fake.next;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, ListNode*, ListNode*>> CASES = {
        {of({1, 2, 4}), of({1, 3, 4}), of({1, 1, 2, 3, 4, 4})},
        {of({}), of({}), of({})},
        {of({}), of({0}), of({0})},
    };

    for (auto& [list1, list2, excepted] : CASES) {
        assert(eq(o.mergeTwoLists(list1, list2), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.mergeTwoLists(of({1, 2, 4}), of({1, 3, 4}));
        o.mergeTwoLists(of({}), of({}));
        o.mergeTwoLists(of({}), of({0}));
    }

    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0021.py
from typing import Optional

from structs.ListNode import ListNode, create


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        fake = walk = ListNode()
        while list1 and list2:
            if list1.val <= list2.val:
                walk.next = list1
                list1 = list1.next
            else:
                walk.next = list2
                list2 = list2.next
            walk = walk.next
        walk.next = list1 or list2
        return fake.next


if __name__ == '__main__':
    import unittest

    CASES = (
        (create(1, 2, 4), create(1, 3, 4), create(1, 1, 2, 3, 4, 4)),
        (None, None, None),
        (None, ListNode(0), ListNode(0)),
    )

    class Test(unittest.TestCase):
        def test(self):
            mergeTwoLists = Solution().mergeTwoLists
            for l1, l2, excepted in CASES:
                actual = mergeTwoLists(l1, l2)
                self.assertEqual(actual, excepted)

    unittest.main()
